### PR TITLE
Modify upgrade scripts to obtain correct server level collation during upgrade

### PR DESCRIPTION
**Pull request by ahmed-shameem**
### Description
In all MVU sql scripts, we update the server collation name and reset the restored_server_collation_name GUC. During the upgrade from 16.1 --> latest (17), the upgrade started failing for non-default server collation like chinese_prc_ci_as as we did not have correct collation information and the collation of columns of views were trying to get updated. This is because, we never updated the collation related information in any of the upgrade script post 16.1, hence it was always picking the default value which is bbf_unicode_cp1_ci_as. 
The issue may come up in case when user is performing upgrade from some version of 15 (before 15.6) to latest if we have defines some view which contains collatable column in upgrade script and the server level collation is different than default collation.

We address the issue by updating the collation related information in the necessary upgrade scripts. 

### Issues Resolved ###
BABEL-5415
Signed-off-by: Shameem Ahmed [shmeeh@amazon.com](mailto:shmeeh@amazon.com)

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.4.0--4.0.0.sql [hunk 1]
```diff
@@ -1,2 +1,2 @@
 -- complain if script is sourced in psql, rather than via ALTER EXTENSION
-\echo Use "ALTER EXTENSION ""babelfishpg_common"" UPDATE TO '4.0.0'" to load this file. \quit
+\echo Use "ALTER EXTENSION ""babelfishpg_common"" UPDATE TO '4.0.0'" to load this file. \quit
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.4.0--4.0.0.sql]: Unnecessary change.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql [hunk 2]
```diff
@@ -31,6 +31,26 @@ end
...
+do
+language plpgsql
+$$
+    declare
+        query text;
+    begin
+        query := pg_catalog.format('alter database %s reset babelfishpg_tsql.restored_server_collation_name', CURRENT_DATABASE());
+        execute query;
+    end;
+$$;
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:62]: I am thinking if we re-set this guc once in the first upgrade script then what happens when we again execute the above procedure in the next upgrade script with the default value (NULL?). Are we handling this in the `C` implementation of the procedure?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql [hunk 3]
```diff
@@ -31,6 +31,26 @@ end
 $$
 LANGUAGE plpgsql;
 
+CREATE OR REPLACE FUNCTION sys.babelfish_update_server_collation_name() RETURNS VOID
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:34]: if this issue only comes up when we upgrade from 15.x (x<=6) to 17 latest then how is this upgrade path coming into picture?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql [hunk 4]
```diff
@@ -31,6 +31,26 @@ end
 $$
 LANGUAGE plpgsql;
 
+CREATE OR REPLACE FUNCTION sys.babelfish_update_server_collation_name() RETURNS VOID
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql:34]: I am not sure why is this required?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.1.0--4.2.0.sql [hunk 5]
```diff
@@ -31,6 +31,26 @@ end
 $$
 LANGUAGE plpgsql;
 
+CREATE OR REPLACE FUNCTION sys.babelfish_update_server_collation_name() RETURNS VOID
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.1.0--4.2.0.sql:34]: this also should not be required. only 4.0--4.1 is needed to handle upgrades

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.4.0--5.0.0.sql [hunk 6]
```diff
@@ -37,6 +37,27 @@ LANGUAGE plpgsql;
  * So make sure that any SQL statement (DDL/DML) being added here can be executed multiple times without affecting
  * final behaviour.
  */
+
+CREATE OR REPLACE FUNCTION sys.babelfish_update_server_collation_name() RETURNS VOID
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.4.0--5.0.0.sql:41]: this should also be not required

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql (hunk 3 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:34]: It is possible that when source is 15.x (x <6) and targrt is 17.y, then we will come till 15.6 and then upgrade to 16 and then 17. For the first step, it is possible that the collation of some view may have been changed. As our upgrade test framework does not cover all such possible upgrade paths for non-default collations, added this to be on the safe side.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql (hunk 4 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql:34]: Same reason as above. We need to put update the collation info for all such versions considering the user is on 15.1, 15.2 and so on. Only one upgrade script is not sufficient to address the issue.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.1.0--4.2.0.sql (hunk 5 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.1.0--4.2.0.sql:34]: Actually, if a user is on extension version 4.3/4.4 (and so on) and trying to go to 5.x then they may face the same error.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.4.0--5.0.0.sql (hunk 6 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--4.4.0--5.0.0.sql:41]: This is required because if any user is on extension version 4.4 and trying to upgrade to 5.0, the upgrade will fail.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql (hunk 3 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:34]: Got it

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql (hunk 2 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:62]: yes, this will cause an issue as we are raising error if restored_collation_name is NULL

Comment (rishabhtanwar29) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:62]: But that doesn't sound right isn't it? We are going to repeatedly call this procedure and reset this GUC in each upgrade script so does that mean we are going to fail the second time we call this procedure since the GUC would have been reset now?

Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql:62]: Updated in latest rev.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql [hunk 7]
```diff
@@ -31,6 +31,37 @@ end
...
+LANGUAGE C
+AS 'babelfishpg_common', 'babelfish_update_server_collation_name';
+
+DO
+LANGUAGE plpgsql
+$$
+BEGIN
+    -- Check if the GUC is empty
+    IF current_setting('babelfishpg_tsql.restored_server_collation_name', true) <> '' 
+        OR current_setting('babelfishpg_tsql.restored_server_collation_name', true) IS NOT NULL THEN
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql]: This should be `AND` right? Also, we should first check if it is NOT NULL and then check if it not empty.

File: contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.4.0--4.0.0.sql (hunk 1 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_common/sql/upgrades/babelfishpg_common--3.4.0--4.0.0.sql]: Removed in latest rev

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql (hunk 7 above)
Comment (ahmed-shameem) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql]: I have updated the condition in latest rev. As discussed offline, NULL check is not required, check for empty string is sufficient.

**Review by rishabhtanwar29 — APPROVED** [review]
LGTM

**Review by shalinilohia50 — APPROVED** [review]

**Review by Deepesh125 — APPROVED** [review]

**Review by Deepesh125 — APPROVED** [review]
